pkg/controller/cachemanager/cachemanager_test.go

```diff
@@ -0,0 +1,455 @@
+	cm := unstructuredFor(configMapGVK, "config-test-1")
+	require.NoError(t, c.Create(ctx, cm), "creating ConfigMap config-test-1")
+
+	cm2 := unstructuredFor(configMapGVK, "config-test-2")
+	require.NoError(t, c.Create(ctx, cm2), "creating ConfigMap config-test-2")
+
+	pod := unstructuredFor(podGVK, "pod-1")
+	require.NoError(t, c.Create(ctx, pod), "creating Pod pod-1")
+
+	opaClient, ok := cacheManager.opa.(*fakes.FakeOpa)
+	require.True(t, ok)
+
+	syncSourceOne := aggregator.Key{Source: "source_a", ID: "ID_a"}
+	require.NoError(t, cacheManager.AddSource(ctx, []schema.GroupVersionKind{configMapGVK, podGVK}, syncSourceOne.Source, syncSourceOne.ID))
+
+	expected := map[fakes.OpaKey]interface{}{
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+		{Gvk: configMapGVK, Key: "default/config-test-2"}: nil,
+		{Gvk: podGVK, Key: "default/pod-1"}:               nil,
+	}
+
+	require.Eventually(t, expectedCheck(opaClient, expected), testutils.TenSecond, testutils.OneSecond)
+
+	// now assert that the gvkAggregator looks as expected
+	cacheManager.gvkAggregator.IsPresent(configMapGVK)
+	gvks := cacheManager.gvkAggregator.List(syncSourceOne)
+	require.Len(t, gvks, 2)
+	_, foundConfigMap := gvks[configMapGVK]
+	require.True(t, foundConfigMap)
+	_, foundPod := gvks[podGVK]
+	require.True(t, foundPod)
+
+	// now remove the podgvk and make sure we don't have pods in the cache anymore
+	require.NoError(t, cacheManager.AddSource(ctx, []schema.GroupVersionKind{configMapGVK}, syncSourceOne.Source, syncSourceOne.ID))
+
+	expected = map[fakes.OpaKey]interface{}{
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+		{Gvk: configMapGVK, Key: "default/config-test-2"}: nil,
+	}
+	require.Eventually(t, expectedCheck(opaClient, expected), testutils.TenSecond, testutils.OneSecond)
+	// now assert that the gvkAggregator looks as expected
+	cacheManager.gvkAggregator.IsPresent(configMapGVK)
+	gvks = cacheManager.gvkAggregator.List(syncSourceOne)
+	require.Len(t, gvks, 1)
+	_, foundConfigMap = gvks[configMapGVK]
+	require.True(t, foundConfigMap)
+	_, foundPod = gvks[podGVK]
+	require.False(t, foundPod)
+
+	// now make sure that adding another sync source with the same gvk has no side effects
+	syncSourceTwo := aggregator.Key{Source: "source_b", ID: "ID_b"}
+	require.NoError(t, cacheManager.AddSource(ctx, []schema.GroupVersionKind{configMapGVK}, syncSourceTwo.Source, syncSourceTwo.ID))
+
+	reqConditionForAgg := func() bool {
+		cacheManager.gvkAggregator.IsPresent(configMapGVK)
+		gvks := cacheManager.gvkAggregator.List(syncSourceOne)
+		if len(gvks) != 1 {
+			return false
+		}
+		_, found := gvks[configMapGVK]
+		if !found {
+			return false
+		}
+
+		gvks2 := cacheManager.gvkAggregator.List(syncSourceTwo)
+		if len(gvks2) != 1 {
+			return false
+		}
+		_, found2 := gvks2[configMapGVK]
+		if !found2 {
+			return false
+		}
+
+		return true
+	}
+	require.Eventually(t, reqConditionForAgg, testutils.TenSecond, testutils.OneSecond)
```

Comment:
Do we want possibly slow tests like this our unit tests?  If we are wanting to test the real waiting and sequencing, an e2e may be a better fit.

One way to do things like this effectively is to mock the clock in the CacheManager and deterministically create the sequence of events that you want to test.

That said, I'd personally just chop this test up into smaller tests that exercise specific scenarios like:

1. Adding a Source yields that GVK in the cache
2. Removing a source removes that GVK from the cache
3. Adding a source that already exists has no effect
4. Deleting a source that doesn't exist has no effect

If you've got each of those covered, you've got short, clear tests that get you most of the way.  Is there additional value in creating combinations of those?

Reply:
I've moved integration tests to their own file. I've added unit tests for some of the scenarios above too. Let me know what you think!